HLL: Avoid some allocations when possible.

gianm opened the pull request:
- HLLC.fold avoids duplicating the other buffer by saving and restoring its position.
- HLLC.makeCollector(buffer) no longer duplicates incoming BBs.
- Updated call sites where appropriate to duplicate BBs passed to HLLC.

Inspired by @navis's comment in https://github.com/druid-io/druid/pull/3111#issuecomment-236120591 (although I didn't go as far 😄 ).

Benchmarks, jdk1.8.0_60 (19–30% speedup depending on the test):

```
topN - master + sorting on hyperUniquesMet

Benchmark                                  (numSegments)  (rowsPerSegment)        (schemaAndQuery)  (threshold)  Mode  Cnt        Score       Error  Units
TopNBenchmark.queryMultiQueryableIndex                 1            750000                 basic.A           10  avgt   25   193478.532 ±  5680.045  us/op
TopNBenchmark.querySingleIncrementalIndex              1            750000                 basic.A           10  avgt   25  1571360.141 ± 37579.656  us/op
TopNBenchmark.querySingleQueryableIndex                1            750000                 basic.A           10  avgt   25   192961.326 ±  5667.723  us/op

topN - patch + sorting on hyperUniquesMet 

Benchmark                                  (numSegments)  (rowsPerSegment)  (schemaAndQuery)  (threshold)  Mode  Cnt        Score       Error  Units
TopNBenchmark.queryMultiQueryableIndex                 1            750000           basic.A           10  avgt   25   158187.083 ±  4630.679  us/op
TopNBenchmark.querySingleIncrementalIndex              1            750000           basic.A           10  avgt   25  1082888.189 ± 33717.391  us/op
TopNBenchmark.querySingleQueryableIndex                1            750000           basic.A           10  avgt   25   155275.814 ±  4052.757  us/op
```

fjy commented:
👍 

@navis do you mind reviewing this one as well?

drcrallen commented on processing/src/main/java/io/druid/query/aggregation/cardinality/CardinalityBufferAggregator.java:
```diff
@@ -53,16 +53,23 @@ public void init(ByteBuffer buf, int position)
-      CardinalityAggregator.hashRow(selectorList, collector);
-    } else {
-      CardinalityAggregator.hashValues(selectorList, collector);
+    // Save position, limit and restore later instead of allocating a new ByteBuffer
```
This shouldn't be *allocating* a new byte buffer. `duplicate()` simply makes a new buffer view into the underlying data. Are you sure this one affects performance?

drcrallen replied:
Actually, I believe it. Even for "small" allocations, since this is in `aggregate` it is going to be called again and again and again.

gianm replied:
I didn't test Cardinality, but I did test HyperUnique, and the performance boost from avoiding the `duplicate()` of the left-hand side is substantial (a significant chunk of the overall ~19% boost).

gianm replied:
btw, by "allocating a new ByteBuffer" I really meant "allocating a new ByteBuffer object pointing to the existing data". The ByteBuffer object has substantial overhead (8 fields!) and apparently the JVM has trouble optimizing that even for short lived allocations.

fwiw- I also tried adjusting things such that the HyperLogLogCollector was re-used too, but saw zero performance boost from that. I thought that was cool, and figured that was because it only has 3 fields, which made is easier for the JVM to optimize.

drcrallen replied:
Can you modify the comment to make a clearer distinction between bytebuffer object allocation and the buffer allocation itself?

gianm replied:
yes, definitely.

gianm replied:
Updated the comment to say "ByteBuffer object" instead of "ByteBuffer"

drcrallen commented on processing/src/main/java/io/druid/query/aggregation/hyperloglog/HyperLogLogCollector.java:
```diff
@@ -92,8 +92,23 @@ public static HyperLogLogCollector makeLatestCollector()
+   * Create a wrapper object around an HLL sketch contained within a buffer.
+   *
+   * @param buffer      buffer containing HLL sketch
+   * @param bufferOwned if true, we are permitted to change the position and limit of the buffer
```
I propose changing the contract here such that ONLY buffers who can have their limit and position changed can be passed.

drcrallen replied:
Then the old method can be deprecated

gianm replied:
What's the advantage of that over the boolean?

drcrallen replied:
Readability / maintainability. From a program flow standpoint I don't think it makes a difference, but from any new reviewer seeing `HyperLogLogCollector.of(buffer1.duplicate())` and `HyperLogLogCollector.of(buffer2)` with the javadoc signature clarifying that the buffer can have its limit and position modified IMHO is easier to grok than `HyperLogLogCollector.makeCollector(buffer1, false)` vs `HyperLogLogCollector.makeCollector(buffer2, true)`

drcrallen replied:
or `HyperLogLogCollector.wrap(buffer1.duplicate())` and `HyperLogLogCollector.wrap(buffer2)`

gianm replied:
How do you feel about changing the contract of `makeCollector` to assume the buffer is owned (i.e. callers need to duplicate if they want to continue using the buffer)? I think having two methods is confusing, and I can just update all the call sites.

drcrallen replied:
Mentioned in private conversation:

The only reservation I would have is that there could be extensions which make the assumption of immutability. But as @gianm mentioned, having extensions call APIs that are not in druid-api is risky business. So I'm good with the proposed mutability change

gianm commented:
@drcrallen I updated the `makeCollector` contract to say it's allowed to modify the position/limit of the incoming buffer, and updated call sites accordingly

drcrallen commented on processing/src/main/java/io/druid/query/aggregation/hyperloglog/HyperLogLogCollector.java:
```diff
@@ -353,65 +369,74 @@ public HyperLogLogCollector fold(HyperLogLogCollector other)
 
-    byte myOffset = getRegisterOffset();
-    short numNonZero = getNumNonZeroRegisters();
+    // Save position and restore later to avoid allocations due to duplicating the buffer
```
object

gianm replied:
updated

drcrallen commented:
Minor doc comment at https://github.com/druid-io/druid/pull/3314/files#r73394157 but 👍 after doc change and travis

navis commented:
@gianm Looks good. I think I should have dig in more on that.
Just one thing : isn't it possible to lessen overhead of "swap" part in fold() operation in HyperLogLogCollector?

gianm commented:
@navis possibly, to be honest I did not look that closely, and just made the most obvious modifications.

drcrallen commented:
I support small incremental improvement/changes.

navis commented:
ok, I think I can do that after this. 👍 from me.
